research-app/src/App.vue

```diff
@@ -1213,6 +1361,369 @@ export default class App extends WWTAwareComponent {
+        const settings = state.settings;
+        const values: any[] = imageSetLayerSettingNames.map(setting => (settings as any)["get_" + setting]());
+        imagerySettingMessages.push({
+          event: "image_layer_set_multi",
+          id: imageset.get_name(),
+          settings: imageSetLayerSettingNames,
+          values: values,
+        });
+
+        imageryStretchMessages.push({
+          event: "image_layer_stretch",
+          id: imageset.get_name(),
+          version: 1, // Don't think that this should matter - we're only sending one per layer
+          stretch: state.scaleType,
+          vmin: state.vmin,
+          vmax: state.vmax,
+        });
+      }
+    });
+
+    const sourceMessages: selections.AddSourceMessage[] = this.sources.map(source => {
+      return {
+        type: "add_source",
+        source: this.prepareForMessaging(source),
+      };
+    });
+
+    const createAnnotationMessages: classicPywwt.CreateAnnotationMessage[] = [];
+    const annotationSettingsMessages: layers.MultiModifyAnnotationMessage[] = [];
+    for (const [id, handler] of this.annotations) {
+      const annotation = handler.annotation();
+      let shape: "circle" | "line" | "polygon" | undefined = undefined;
+      if (annotation instanceof Circle) {
+        shape = "circle";
+      } else if (annotation instanceof Poly) {
+        shape = "polygon";
+      } else if (annotation instanceof PolyLine) {
+        shape = "line"
+      }
+      if (shape == undefined) {
+        continue;
+      }
+      createAnnotationMessages.push({
+        event: "annotation_create",
+        shape: shape,
+        id: id,
+      });
+
+      let settingNames: string[] = [];
+      if (shape === "circle") {
+        settingNames = circleAnnotationSettingNames;
+      } else if (shape === "polygon") {
+        settingNames = polyAnnotationSettingNames;
+      } else {
+        settingNames = polyLineAnnotationSettingNames;
+      }
+      const values = settingNames.map(setting => (annotation as any)["get_" + setting]());
+      annotationSettingsMessages.push({
+        event: "annotation_set_multi",
+        id: id,
+        settings: settingNames,
+        values: values,
+      });
+    }
+
+    const messageStrings = [
+      coordinatesMessage,
+      backgroundMessage,
+      foregroundMessage,
+      ...loadCatalogsMessages,
+      ...catalogSettingsMessages,
+      ...loadWtmlMessages,
+      ...imageryLayerMessages,
+      ...imagerySettingMessages,
+      ...imageryStretchMessages,
+      ...sourceMessages,
+      ...createAnnotationMessages,
+      ...annotationSettingsMessages,
+    ].flatMap(s => s ? [this.encodeObjectBase64(s)] : []);
+
+    const messageString = messageStrings.join(",");
+    const outString = messageString;
+
+    const params = {
+      origin: window.location.origin,
+      messages: outString,
+    };
+
+    return window.location.origin + '/?' + (new URLSearchParams(params)).toString();
```

Comment:
This isn't generically correct since `window.location.origin` strips off the URL path of wherever the app instance is being hosted. I think the proper way to do this is copy `window.location` as a parsed-out URL, overwrite its query string, and then stringify the resulting URL.

Reply:
I did what you suggested here - I created a `URL` object from the window location, overwrote its `search` member, and obtained the string form of the new URL.